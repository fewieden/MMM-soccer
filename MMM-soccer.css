.MMM-soccer .icon {
    width: 24px;
    height: 24px;
}

.MMM-soccer .no-color {
    -webkit-filter: grayscale(100%);
}

.MMM-soccer .table {
    border-spacing: 2px;
}

.MMM-soccer .centered-row {
    vertical-align: middle;
}

.MMM-soccer .space-between {
    display: flex;
    justify-content: space-between;
}

.MMM-soccer .align-left {
    text-align: left;
}

.MMM-soccer .align-right {
    text-align: right;
}

.MMM-soccer .centered {
    text-align: center;
}
