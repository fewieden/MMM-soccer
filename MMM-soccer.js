/**
 * @file MMM-soccer.js
 *
 * @author fewieden
 * @license MIT
 *
 * @see  https://github.com/fewieden/MMM-soccer
 */

/**
 * @external Module
 * @see https://github.com/MagicMirrorOrg/MagicMirror/blob/master/js/module.js
 */

/**
 * @external Log
 * @see https://github.com/MagicMirrorOrg/MagicMirror/blob/master/js/logger.js
 */

/**
 * @module MMM-soccer
 * @description Frontend of the MagicMirror² module.
 *
 * @requires external:Module
 * @requires external:Log
 */
Module.register('MMM-soccer', {
    /**
     * @member {string} requiresVersion - Defines the required minimum version of the MagicMirror framework in order to
     * run this version of the module.
     */
    requiresVersion: '2.14.0',

    /**
     * @member {Object} defaults - Defines the default config values.
     * @property {boolean|string} api_key - API acces key for football-data.org.
     * @property {boolean} colored - Flag to show logos in color or black/white.
     * @property {string} show - Country name (uppercase) to be shown in module.
     * @property {boolean|Object} focus_on - Hash of country name -> club name to determine highlighted team per league.
     * @property {boolean|int} max_teams - Maximium amount of teams to be displayed.
     * @property {boolean} logos - Flag to show club logos.
     * @property {Object} leagues - Hash of country name -> league id.
     */
    defaults: {
        api_key: false,
        colored: false,
        show: 'GERMANY',
        focus_on: false,
        max_teams: false,
        logos: false,
        leagues: {
            GERMANY: 'BL1',
            FRANCE: 'FL1',
            ENGLAND: 'PL',
            SPAIN: 'PD',
            ITALY: 'SA'
        }
    },

    /**
     * @member {Object} voice - Defines the default mode and commands of this module.
     * @property {string} mode - Voice mode of this module.
     * @property {string[]} sentences - List of voice commands of this module.
     */
    voice: {
        mode: 'SOCCER',
        sentences: [
            'OPEN HELP',
            'CLOSE HELP',
            'SHOW STANDINGS OF COUNTRY NAME',
            'EXPAND VIEW',
            'COLLAPSE VIEW'
        ]
    },

    /**
     * @member {boolean} loading - Flag to indicate the loading state of the module.
     */
    loading: true,
    /**
     * @member {Object[]} standing - Stores the list of standing table entries of current selected league.
     */
    standing: [],
    /**
     * @member {Object} competition - Details about the current selected league.
     */
    competition: {},
    /**
     * @member {Object} season - Details about the current season of the selected league.
     */
    season: {},

    /**
     * @function start
     * @description Adds nunjuck filters and requests for league data.
     * @override
     *
     * @returns {void}
     */
    start() {
        Log.info(`Starting module: ${this.name}`);
        this.addFilters();
        this.currentLeague = this.config.leagues[this.config.show];
        this.getData();
        setInterval(() => {
            this.getData();
        }, this.config.api_key ? 300000 : 1800000); // with api_key every 5min without every 30min
    },

    /**
     * @function getData
     * @description Sends request to the node_helper to fetch data for the current selected league.
     *
     * @returns {void}
     */
    getData() {
        this.sendSocketNotification('GET_DATA', {league: this.currentLeague, api_key: this.config.api_key});
    },

    /**
     * @function socketNotificationReceived
     * @description Handles incoming messages from node_helper.
     * @override
     *
     * @param {string} notification - Notification name
     * @param {*} payload - Detailed payload of the notification.
     */
    socketNotificationReceived(notification, payload) {
        if (notification === 'DATA') {
            this.standing = payload.standings[0].table;
            this.season = payload.season;
            this.competition = payload.competition;
            this.loading = false;
            this.updateDom(300);
        }
    },

    /**
     * @function notificationReceived
     * @description Handles incoming broadcasts from other modules or the MagicMirror² core.
     * @override
     *
     * @param {string} notification - Notification name
     * @param {*} payload - Detailed payload of the notification.
     * @param {Object} sender - Module that sent the notification or undefined for MagicMirror² core.
     */
    notificationReceived(notification, payload, sender) {
        if (notification === 'ALL_MODULES_STARTED') {
            const leagues = Object.keys(this.config.leagues).join(' ');
            this.sendNotification('REGISTER_VOICE_MODULE', {
                mode: this.voice.mode,
                sentences: [...this.voice.sentences, leagues]
            });
        } else if (notification === 'VOICE_SOCCER' && sender.name === 'MMM-voice') {
            this.executeVoiceCommands(payload);
        } else if (notification === 'VOICE_MODE_CHANGED' && sender.name === 'MMM-voice'
            && payload.old === this.voice.mode) {
            this.sendNotification('CLOSE_MODAL');
        }
    },

    /**
     * @function getStyles
     * @description Style dependencies for this module.
     * @override
     *
     * @returns {string[]} List of the style dependency filepaths.
     */
    getStyles() {
        return ['font-awesome.css', 'MMM-soccer.css'];
    },

    /**
     * @function getTranslations
     * @description Translations for this module.
     * @override
     *
     * @returns {Object.<string, string>} Available translations for this module (key: language code, value: filepath).
     */
    getTranslations() {
        return {
            en: 'translations/en.json',
            de: 'translations/de.json',
            fr: 'translations/fr.json',
            id: 'translations/id.json',
            sv: 'translations/sv.json'
        };
    },

    /**
     * @function getTemplate
     * @description Nunjuck template.
     * @override
     *
     * @returns {string} Path to nunjuck template.
     */
    getTemplate() {
        return 'templates/MMM-soccer.njk';
    },

    /**
     * @function getTemplateData
     * @description Data that gets rendered in the nunjuck template.
     * @override
     *
     * @returns {string} Data for the nunjuck template.
     */
    getTemplateData() {
        return {
            boundaries: this.calculateTeamDisplayBoundaries(),
            competitionName: this.competition.name || this.name,
            config: this.config,
            matchDayNumber: this.season.currentMatchday || 'N/A',
            standing: this.standing,
            loading: this.loading
        };
    },

    /**
     * @function handleHelpModal
     * @description Opens/closes help modal based on voice commands.
     *
     * @param {string} data - Text with commands.
     *
     * @returns {void}
     */
    handleHelpModal(data) {
        if (/(CLOSE)/g.test(data) && !/(OPEN)/g.test(data)) {
            this.sendNotification('CLOSE_MODAL');
        } else if (/(OPEN)/g.test(data) && !/(CLOSE)/g.test(data)) {
            this.sendNotification('OPEN_MODAL', {
                template: 'templates/HelpModal.njk',
                data: {
                    ...this.voice,
                    fns: {
                        translate: this.translate.bind(this)
                    }
                }
            });
        }
    },

    /**
     * @function handleStandingsModal
     * @description Opens/closes standing modal based on voice commands.
     *
     * @param {string} data - Text with commands.
     *
     * @returns {void}
     */
    handleStandingsModal(data) {
        if (/(COLLAPSE)/g.test(data) && !/(EXPAND)/g.test(data)) {
            this.sendNotification('CLOSE_MODAL');
        } else if (/(EXPAND)/g.test(data) && !/(COLLAPSE)/g.test(data)) {
            this.sendNotification('OPEN_MODAL', {
                template: 'templates/StandingsModal.njk',
                data: {
                    ...this.getTemplateData(),
                    fns: {
                        translate: this.translate.bind(this)
                    }
                }
            });
        }
    },

    /**
     * @function handleLeagueSwitch
     * @description Sitches the soccer league based on voice commands.
     *
     * @param {string} data - Text with commands.
     *
     * @returns {void}
     */
    handleLeagueSwitch(data) {
        const countries = Object.keys(this.config.leagues);

        for (const country of countries) {
            const regexp = new RegExp(country, 'g');

            if (regexp.test(data)) {
                this.sendNotification('CLOSE_MODAL');

                if (this.currentLeague !== this.config.leagues[country]) {
                    this.currentLeague = this.config.leagues[country];
                    this.getData();
                }

                break;
            }
        }

        this.updateDom(300);
    },

    /**
     * @function executeVoiceCommands
     * @description Executes the voice commands.
     *
     * @param {string} data - Text with commands.
     *
     * @returns {void}
     */
    executeVoiceCommands(data) {
        if (/(HELP)/g.test(data)) {
            this.handleHelpModal(data);
        } else if (/(VIEW)/g.test(data)) {
            this.handleStandingsModal(data);
        } else if (/(STANDINGS)/g.test(data)) {
            this.handleLeagueSwitch(data);
        }
    },

    /**
     * @function getMaxTeams
     * @description Are there less entries than the config option specifies.
     *
     * @returns {number} Amount of teams to display
     */
    getMaxTeams() {
        if (this.config.max_teams) {
            return Math.min(Math.max(this.config.max_teams, 0), this.standing.length);
        }

        return this.standing.length;
    },

    /**
     * @function findFocusTeam
     * @description Helper function to find index of team in standings
     *
     * @returns {Object} Index of team, first and last team to display. focusTeamIndex is -1 if it can't be found.
     */
    findFocusTeam() {
        let focusTeamIndex = -1;

        for (let i = 0; i < this.standing.length; i += 1) {
            if (this.standing[i].team.name === this.config.focus_on[this.config.show]) {
                focusTeamIndex = i;
                break;
            }
        }

        const {firstTeam, lastTeam} = this.getFirstAndLastTeam(focusTeamIndex);

        return {focusTeamIndex, firstTeam, lastTeam};
    },

    /**
     * @function getFirstAndLastTeam
     * @description Helper function to get the boundaries of the teams that should be displayed.
     *
     * @param {number} index - Index of the focus_on team.
     *
     * @returns {Object} Index of the first and the last team.
     */
    getFirstAndLastTeam(index) {
        let firstTeam = 0;
        let lastTeam = this.standing.length - 1;

        if (this.config.max_teams) {
            const before = parseInt(this.config.max_teams / 2);
            const indexDiff = this.config.max_teams - 1;
            firstTeam = Math.max(index - before, 0);
            if (firstTeam + indexDiff < this.standing.length) {
                lastTeam = firstTeam + indexDiff;
            } else {
                firstTeam = Math.max(lastTeam - indexDiff, 0);
            }
        }

        return {firstTeam, lastTeam};
    },

    /**
     * @function calculateTeamDisplayBoundaries
     * @description Calculates the boundaries of teams based on the config.
     *
     * @returns {Object} Index of team, first and last team to display.
     */
    calculateTeamDisplayBoundaries() {
        if (this.config.focus_on && Object.prototype.hasOwnProperty.call(this.config.focus_on, this.config.show)) {
            if (this.config.focus_on[this.config.show] === 'TOP') {
                return {
                    focusTeamIndex: -1,
                    firstTeam: 0,
                    lastTeam: this.getMaxTeams() - 1
                };
            } else if (this.config.focus_on[this.config.show] === 'BOTTOM') {
                return {
                    focusTeamIndex: -1,
                    firstTeam: this.standing.length - this.getMaxTeams(),
                    lastTeam: this.standing.length - 1
                };
            }
        }

        return this.findFocusTeam();
    },

    /**
     * @function addFilters
     * @description Adds the filter used by the nunjuck template.
     *
     * @returns {void}
     */
    addFilters() {
        this.nunjucksEnvironment().addFilter('fade', (index, focus) => {
            if (this.config.max_teams && focus >= 0) {
                if (index !== focus) {
                    const currentStep = Math.abs(index - focus);
                    const percentage = (1 - 1 / this.config.max_teams * currentStep).toFixed(2);

                    return `opacity: ${percentage}`;
                }
            }

            return '';
        });
    }
});
