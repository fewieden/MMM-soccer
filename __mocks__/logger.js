const logger = {
    debug: jest.fn(),
    log: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
};

global.Log = logger;

module.exports = logger;
