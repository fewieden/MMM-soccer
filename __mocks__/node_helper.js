const Log = require("logger");

module.exports = {
    create(overrides) {
        const base = {
            init() {
                Log.log("Initializing new module helper ...");
            },

            loaded(callback) {
                Log.log(`Module helper loaded: ${this.name}`);
                callback();
            },

            start() {
                Log.log(`Starting module helper: ${this.name}`);
            },

            stop() {
                Log.log(`Stopping module helper: ${this.name}`);
            },

            socketNotificationReceived(notification, payload) {
                Log.log(`${this.name} received a socket notification: ${notification} - Payload: ${payload}`);
            },

            setName(name) {
                this.name = name;
            },

            setPath(path) {
                this.path = path;
            },

            sendSocketNotification: jest.fn(),

            setExpressApp: jest.fn(),

            setSocketIO: jest.fn()
        };

        return {...base, ...overrides};
    }
};
