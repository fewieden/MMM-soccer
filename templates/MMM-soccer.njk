{% if loading %}
    <div class="dimmed light">{{ "LOADING" | translate | safe }}</div>
{% else %}
    <header class="space-between">
        <span>{{ competitionName }}</span>
        <span>{{ "MATCHDAY" | translate({matchDayNumber: matchDayNumber}) }}</span>
    </header>
    <table class="xsmall table">
        <thead>
            <tr class="row">
                <th></th>
                {% if config.logos %}
                    <th></th>
                {% endif %}
                <th class="name">{{ "TEAM" | translate }}</th>
                <th class="centered"><i class="fa fa-line-chart"/></th>
                <th class="centered"><i class="fa fa-soccer-ball-o"/></th>
            </tr>
        </thead>
        <tbody>
            {% for index in range(boundaries.firstTeam, boundaries.lastTeam + 1) %}
                <tr class="centered-row {% if index === boundaries.focusTeamIndex %}bright{% endif %}" style="{{ index | fade(boundaries.focusTeamIndex) }}">
                    <td>{{ standing[index].position }}</td>
                    {% if config.logos %}
                        <td><img class="icon {% if not config.colored %}no-color{% endif %}" src="{{ standing[index].team.crest }}"/></td>
                    {% endif %}
                    <td class="name">{{ standing[index].team.name }}</td>
                    <td class="centered">{{ standing[index].points }}</td>
                    <td class="centered">{{ standing[index].goalDifference }}</td>
                </tr>
            {% endfor %}
        </tbody>
    </table>
{% endif %}
