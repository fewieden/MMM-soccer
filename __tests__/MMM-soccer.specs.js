const {readdirSync} = require('node:fs');
const {join} = require('node:path');

require('../__mocks__/logger');
require('../__mocks__/Module');
const {generateResponse} = require('../__mocks__/mockResponse');

describe('MMM-soccer', () => {
    const name = 'MMM-soccer';

    let MMMsoccer;

    beforeEach(() => {
        jest.isolateModules(() => {
            require('../MMM-soccer');
        });

        MMMsoccer = global.Module.create(name);
        MMMsoccer.setData({name, identifier: `Module_1_${name}`});
    });

    test('requires version 2.14', () => {
        expect(MMMsoccer.requiresVersion).toBe('2.14.0');
    });

    test('has correct defaults', () => {
        expect(MMMsoccer.defaults).toMatchSnapshot();
    });

    test('has correct voice commands', () => {
        expect(MMMsoccer.voice).toMatchSnapshot();
    });

    test('inits module with state loading === true', () => {
        expect(MMMsoccer.loading).toBe(true);
    });

    describe('start', () => {
        beforeEach(() => {
            jest.useFakeTimers();
        });

        afterEach(() => {
            jest.useRealTimers();
        });

        test('logs start of module', () => {
            MMMsoccer.start();

            expect(global.Log.info).toHaveBeenNthCalledWith(1, 'Starting module: MMM-soccer');
        });

        test('adds nunjuck filters', () => {
            MMMsoccer.start();

            const addFilter = MMMsoccer.nunjucksEnvironment().addFilter;
            expect(addFilter).toBeCalledTimes(1);
            expect(addFilter).toHaveBeenNthCalledWith(1, 'fade', expect.any(Function));
        });

        test('inits current league', () => {
            MMMsoccer.start();

            expect(MMMsoccer.currentLeague).toBe('BL1');
        });

        test('requests data from node_helper', () => {
            MMMsoccer.start();

            expect(MMMsoccer.sendSocketNotification).toHaveBeenNthCalledWith(1, 'GET_DATA', {
                api_key: false, league: 'BL1'
            });
        });

        test('interval requests data from node_helper', () => {
            MMMsoccer.start();

            jest.runOnlyPendingTimers();

            expect(MMMsoccer.sendSocketNotification).toHaveBeenCalledTimes(2);
            expect(MMMsoccer.sendSocketNotification).toHaveBeenNthCalledWith(2, 'GET_DATA', {
                api_key: false, league: 'BL1'
            });
        });

        test('inits interval correctly for usage with api_key', () => {
            const intervalSpy = jest.spyOn(global, 'setInterval');

            MMMsoccer.start();

            expect(intervalSpy).toHaveBeenNthCalledWith(1, expect.any(Function), 1800000);

            intervalSpy.mockRestore();
        });

        test('inits interval correctly for usage without api_key', () => {
            const intervalSpy = jest.spyOn(global, 'setInterval');

            MMMsoccer.setConfig({api_key: 'TEST_API_KEY'});

            MMMsoccer.start();

            expect(intervalSpy).toHaveBeenNthCalledWith(1, expect.any(Function), 300000);

            intervalSpy.mockRestore();
        });
    });

    describe('socketNotificationReceived', () => {
        const payload = generateResponse();

        test('sets loading to false after receiving data', () => {
            MMMsoccer.socketNotificationReceived('DATA', payload);

            expect(MMMsoccer.loading).toBe(false);
        });

        test('updates dom after receiving data', () => {
            MMMsoccer.socketNotificationReceived('DATA', payload);

            expect(MMMsoccer.updateDom).toHaveBeenNthCalledWith(1, 300);
        });

        test('assigns data correctly', () => {
            MMMsoccer.socketNotificationReceived('DATA', payload);

            expect(MMMsoccer.standing).toMatchObject(payload.standings[0].table);
            expect(MMMsoccer.competition).toMatchObject(payload.competition);
            expect(MMMsoccer.season).toMatchObject(payload.season);
        });

        test('does nothing if notification is different than DATA', () => {
            MMMsoccer.socketNotificationReceived('OTHER NOTIFICATION', payload);

            expect(MMMsoccer.loading).toBe(true);
        });
    });

    describe('notificationReceived', () => {
        test('registers voice commands', () => {
            MMMsoccer.notificationReceived('ALL_MODULES_STARTED');

            expect(MMMsoccer.sendNotification.mock.calls[0]).toMatchSnapshot();
        });

        test('executes voice commands', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'OPEN HELP', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification.mock.calls[0]).toMatchSnapshot();
        });

        test('closes modal after current voice mode was changed to other module', () => {
            MMMsoccer.notificationReceived('VOICE_MODE_CHANGED', {old: 'SOCCER'}, {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'CLOSE_MODAL');
        });

        test('does NOT trigger notifications', () => {
            MMMsoccer.notificationReceived('OTHER NOTIFICATION');

            expect(MMMsoccer.sendNotification).not.toHaveBeenCalled();
        });
    });

    test('registers font-awesome and custom css', () => {
        expect(MMMsoccer.getStyles()).toMatchObject(['font-awesome.css', 'MMM-soccer.css']);
    });

    test('translations are matching translation files', () => {
        const translations = MMMsoccer.getTranslations();

        const translationFiles = readdirSync(join(__dirname, '..', 'translations'));

        expect(Object.keys(translations).length).toBe(5);
        expect(translationFiles.length).toBe(5);

        for (const file of translationFiles) {
            const language = file.replace('.json', '');
            expect(translations[language]).toBe(`translations/${language}.json`);
        }
    });

    test('returns template path inside templates dir', () => {
        expect(MMMsoccer.getTemplate()).toBe(`templates/${name}.njk`);
    });

    describe('executeVoiceCommands', () => {
        test('opens help modal', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'OPEN HELP', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'OPEN_MODAL', {
                data: expect.any(Object),
                template: 'templates/HelpModal.njk'
            });
        });

        test('closes help modal', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'CLOSE HELP', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'CLOSE_MODAL');
        });

        test('does NOTHING if help modal but NO open/close keyword', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'HELP', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).not.toHaveBeenCalled();
        });

        test('opens standings modal', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'EXPAND VIEW', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'OPEN_MODAL', {
                data: expect.any(Object),
                template: 'templates/StandingsModal.njk'
            });
        });

        test('closes standings modal', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'COLLAPSE VIEW', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'CLOSE_MODAL');
        });

        test('does NOTHING if standings modal but NO expand/collapse keyword', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'VIEW', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).not.toHaveBeenCalled();
        });

        test('switches current league to premiere league and requests new data', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'STANDINGS ENGLAND', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'CLOSE_MODAL');
            expect(MMMsoccer.currentLeague).toBe('PL');
            expect(MMMsoccer.sendSocketNotification).toHaveBeenNthCalledWith(1, 'GET_DATA', {
                api_key: false, league: 'PL'
            });
        });

        test('requests NO data if user tries to switch to same league', () => {
            MMMsoccer.currentLeague = 'BL1';
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'STANDINGS GERMANY', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).toHaveBeenNthCalledWith(1, 'CLOSE_MODAL');
            expect(MMMsoccer.sendSocketNotification).not.toHaveBeenCalled();
        });

        test('does NOT switch league if league/country is unknown', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'VIEW', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).not.toHaveBeenCalled();
            expect(MMMsoccer.sendSocketNotification).not.toHaveBeenCalled();
        });

        test('does NOTHING for other voice commands', () => {
            MMMsoccer.notificationReceived('VOICE_SOCCER', 'OTHER COMMAND', {name: 'MMM-voice'});

            expect(MMMsoccer.sendNotification).not.toHaveBeenCalled();
            expect(MMMsoccer.sendSocketNotification).not.toHaveBeenCalled();
        });
    });

    describe('calculateTeamDisplayBoundaries', () => {
        beforeEach(() => {
            MMMsoccer.standing = [
                {team: {name: 'team01'}},
                {team: {name: 'team02'}},
                {team: {name: 'team03'}},
                {team: {name: 'team04'}},
                {team: {name: 'team05'}},
                {team: {name: 'team06'}},
                {team: {name: 'team07'}},
                {team: {name: 'team08'}},
                {team: {name: 'team09'}},
                {team: {name: 'team10'}}
            ];
        });

        test('returns all teams and focus === -1 if focus NOT specified WITHOUT max_teams', () => {
            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 0,
                lastTeam: 9
            });
        });

        test('returns all teams and focus === -1 if focus NOT found WITHOUT max_teams', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'anotherTeam'}});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 0,
                lastTeam: 9
            });
        });

        test('returns all teams and focus === -1 if focus top WITHOUT max_teams', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'TOP'}});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 0,
                lastTeam: 9
            });
        });

        test('returns all teams and focus === -1 if focus bottom WITHOUT max_teams', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'BOTTOM'}});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 0,
                lastTeam: 9
            });
        });

        test('returns 5 teams and index of focus team', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'team06'}, max_teams: 5});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: 5,
                firstTeam: 3,
                lastTeam: 7
            });
        });

        test('returns top 5 teams and focus === -1', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'TOP'}, max_teams: 5});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 0,
                lastTeam: 4
            });
        });

        test('returns top 5 teams and focus === -1 if focus NOT found', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'anotherTeam'}, max_teams: 5});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 0,
                lastTeam: 4
            });
        });

        test('returns top 5 teams and focus === 1 if focus is to close to top', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'team02'}, max_teams: 5});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: 1,
                firstTeam: 0,
                lastTeam: 4
            });
        });

        test('returns bottom 5 teams and focus === -1', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'BOTTOM'}, max_teams: 5});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: -1,
                firstTeam: 5,
                lastTeam: 9
            });
        });

        test('returns bottom 5 teams and focus === 8 if focus is to close to bottom', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'team09'}, max_teams: 5});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: 8,
                firstTeam: 5,
                lastTeam: 9
            });
        });

        test('returns 6 teams and index of focus team', () => {
            MMMsoccer.setConfig({focus_on: {GERMANY: 'team06'}, max_teams: 6});

            expect(MMMsoccer.calculateTeamDisplayBoundaries()).toMatchObject({
                focusTeamIndex: 5,
                firstTeam: 2,
                lastTeam: 7
            });
        });
    });

    describe('addFilters', () => {
        let addFilter;

        beforeEach(() => {
            MMMsoccer.addFilters();
            addFilter = MMMsoccer.nunjucksEnvironment().addFilter;
        });

        describe('fade', () => {
            let fadeFilter;

            beforeEach(() => {
                fadeFilter = MMMsoccer.nunjucksEnvironment().addFilter.mock.calls[0][1];
            });

            test('adds fade filters', () => {
                expect(addFilter).toHaveBeenNthCalledWith(1, 'fade', expect.any(Function));
            });

            test('returns opacity style', () => {
                MMMsoccer.setConfig({max_teams: 7});

                expect(fadeFilter(1, 3)).toBe('opacity: 0.71');
            });

            test('returns empty string if max_teams is NOT set', () => {
                expect(fadeFilter(3, -1)).toBe('');
            });

            test('returns empty string if focus team NOT found', () => {
                MMMsoccer.setConfig({max_teams: 7});

                expect(fadeFilter(3, -1)).toBe('');
            });

            test('returns empty string if focus === index', () => {
                MMMsoccer.setConfig({max_teams: 7});

                expect(fadeFilter(3, 3)).toBe('');
            });
        });
    });
});
