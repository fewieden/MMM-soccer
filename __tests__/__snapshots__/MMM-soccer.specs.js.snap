// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MMM-soccer has correct defaults 1`] = `
{
  "api_key": false,
  "colored": false,
  "focus_on": false,
  "leagues": {
    "ENGLAND": "PL",
    "FRANCE": "FL1",
    "GERMANY": "BL1",
    "ITALY": "SA",
    "SPAIN": "PD",
  },
  "logos": false,
  "max_teams": false,
  "show": "GERMANY",
}
`;

exports[`MMM-soccer has correct voice commands 1`] = `
{
  "mode": "SOCCER",
  "sentences": [
    "OPEN HELP",
    "CLOSE HELP",
    "SHOW STANDINGS OF COUNTRY NAME",
    "EXPAND VIEW",
    "COLLAPSE VIEW",
  ],
}
`;

exports[`MMM-soccer notificationReceived executes voice commands 1`] = `
[
  "OPEN_MODAL",
  {
    "data": {
      "fns": {
        "translate": [Function],
      },
      "mode": "SOCCER",
      "sentences": [
        "OPEN HELP",
        "CLOSE HELP",
        "SHOW STANDINGS OF COUNTRY NAME",
        "EXPAND VIEW",
        "COLLAPSE VIEW",
      ],
    },
    "template": "templates/HelpModal.njk",
  },
]
`;

exports[`MMM-soccer notificationReceived registers voice commands 1`] = `
[
  "REGISTER_VOICE_MODULE",
  {
    "mode": "SOCCER",
    "sentences": [
      "OPEN HELP",
      "CLOSE HELP",
      "SHOW STANDINGS OF COUNTRY NAME",
      "EXPAND VIEW",
      "COLLAPSE VIEW",
      "GERMANY FRANCE ENGLAND SPAIN ITALY",
    ],
  },
]
`;
